mechanicalsoup/stateful_browser.py

```diff
@@ -207,14 +207,46 @@ def select_form(self, selector="form", nr=0):
         :return: The selected form as a soup object. It can also be
             retrieved later with the :attr:`form` attribute.
         """
+
+        def find_associated_elements(form_id):
+            """Function to find all associated elements to a form
+                (i.e. an element with a form attribute -> ``form=form_id``)
+            """
+
+            # Elements which can have a form owner
+            form_associated_elements = ["input", "button", "fieldset",
+                                        "object", "output", "select",
+                                        "textarea"]
+
+            found_elements = []
+
+            for element in form_associated_elements:
+                for new_element in self.page.find_all(element,
+                                                      form=form_id):
+                    found_elements.append(new_element)
+
+            return found_elements
+
         if isinstance(selector, bs4.element.Tag):
             if selector.name != "form":
                 raise LinkNotFoundError
+
+            if selector.has_attr('id'):
+                form_id = selector["id"]
+                new_elements = find_associated_elements(form_id)
+                selector.extend(new_elements)
+
             self.__state.form = Form(selector)
         else:
             # nr is a 0-based index for consistency with mechanize
             found_forms = self.page.select(selector,
                                            limit=nr + 1)
+
+            if len(found_forms) > 0 and found_forms[-1].has_attr('id'):
+                form_id = found_forms[-1]["id"]
+                new_elements = find_associated_elements(form_id)
+                found_forms[-1].extend(new_elements)
+
```

Comment:
The changes are looking good, thanks! I have one small suggestion: what if we refactor this a bit to reduce some code duplication? I'm thinking something along the lines of:

```python
if isinstance(selector, bs4.element.Tag):
    ...
    form = selector
else:
    ...
    form = found_forms[-1]

if form.has_attr("id"):
    new_elements = find_associated_elements(form["id"])
    form.extend(new_elements)

self.__state.form = Form(form)

return self.form
```